google/cloud/internal/retry_policy.h

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_RETRY_POLICY_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_RETRY_POLICY_H_
+
+#include "google/cloud/version.h"
+#include <chrono>
+#include <memory>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+/**
+ * Define the interface for retry policies.
+ *
+ * @tparam StatusTypeParam the type used to represent success/failures.
+ * @tparam RetryablePolicyParam the policy to decide if status is retryable.
+ */
+template <typename StatusTypeParam, typename RetryablePolicyParam>
+class RetryPolicy {
+ public:
+  using StatusType = StatusTypeParam;
```

Comment:
Traits are useful for generic code.  They allow you to write:

```C++
template <typename Policy> void MyGenericFunction(Policy&& p) {
  Policy::Status type = blah;
}
```

otherwise you have to write:

```C++
template <template <typename StatusType, typename T> class Policy>
void MyGenericFunction(
  Policy<StatusType, T>&& p) {
  StatusType blah;
}
```

Meh, removed.